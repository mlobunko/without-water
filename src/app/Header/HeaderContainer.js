import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import authorized from 'shared/authorized';
import { HeaderComponent } from './HeaderComponent';

export class HeaderContainer extends React.Component {
  state = {
    isOpen: false
  };

  handleAddRemoveEventListener = () => {
    if (this.state.isOpen) {
      document.addEventListener('click', this.handleDocumentClick);
    } else {
      document.removeEventListener('click', this.handleDocumentClick);
    }
  };

  toggle = () => {
    this.setState(
      {
        isOpen: !this.state.isOpen
      },
      this.handleAddRemoveEventListener
    );
  };

  handleDocumentClick = e => {
    const container = this.element;
    if (e.target !== container && !container.contains(e.target)) {
      this.toggle();
    }
  };

  componentWillUnmount = () => {
    document.removeEventListener('click', this.handleDocumentClick);
  };

  render() {
    return (
      <div ref={element => (this.element = element)}>
        <HeaderComponent
          isOpen={this.state.isOpen}
          logOut={this.props.logOut}
          toggle={this.toggle}
          authorized={this.props.authorized}
          authorizedActionsLogOut={this.props.authorizedActionsLogOut}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  authorized: state.authorized
});

const mapDispatchToProps = {
  authorizedActionsLogOut: authorized.actions.logOut
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(HeaderContainer)
);
