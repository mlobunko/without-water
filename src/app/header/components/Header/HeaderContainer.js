import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

import HeaderComponent from './HeaderComponent';

import authorized from 'shared/authorized';

export class HeaderContainer extends React.Component {
  state = {
    isOpen: false
  };

  handleAddRemoveEventListener = () => {
    if (this.state.isOpen) {
      document.addEventListener('click', this.handleDocumentClick);
    } else {
      document.removeEventListener('click', this.handleDocumentClick);
    }
  };

  toggle = () => {
    this.setState(
      {
        isOpen: !this.state.isOpen
      },
      this.handleAddRemoveEventListener
    );
  };

  handleDocumentClick = e => {
    const container = this.element;
    if (e.target !== container && !container.contains(e.target)) {
      this.toggle();
    }
  };

  componentWillUnmount = () => {
    document.removeEventListener('click', this.handleDocumentClick);
  };

  render() {
    return (
      <div ref={element => (this.element = element)}>
        <HeaderComponent
          isOpen={this.state.isOpen}
          logOut={this.props.logOut}
          toggle={this.toggle}
          authorized={this.props.authorized}
          authorizedActionsLogOut={this.props.authorizedActionsLogOut}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  authorized: authorized.selectors.getAll(state)
});

const mapDispatchToProps = {
  authorizedActionsLogOut: authorized.actions.logOut
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(HeaderContainer)
);

HeaderContainer.propTypes = {
  authorized: PropTypes.bool
};
