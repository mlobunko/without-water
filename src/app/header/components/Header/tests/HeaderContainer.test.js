import React from 'react';
import { shallow } from 'enzyme';

import { HeaderContainer } from '../HeaderContainer';

const props = {
  authorized: true,
  authorizedActionsLogOut: jest.fn()
};

describe('handlers', () => {
  Object.defineProperties(document, {
    addEventListener: {
      value: jest.fn()
    },
    removeEventListener: {
      value: jest.fn()
    }
  });

  describe('handleAddRemoveEventListener method', () => {
    it('handle remove event when isOpen false', () => {
      const initialState = {
        isOpen: false
      };
      const wrapper = shallow(<HeaderContainer {...props} />);
      wrapper.setState(initialState);
      const expectedFunction = wrapper.instance().handleDocumentClick;
      wrapper.instance().handleAddRemoveEventListener();
      expect(document.removeEventListener).toHaveBeenCalledWith(
        'click',
        expectedFunction
      );
    });

    it('handle add event when isOpen true', () => {
      const initialState = {
        isOpen: true
      };
      const wrapper = shallow(<HeaderContainer {...props} />);
      wrapper.setState(initialState);
      const expectedFunction = wrapper.instance().handleDocumentClick;
      wrapper.instance().handleAddRemoveEventListener();
      expect(document.addEventListener).toHaveBeenCalledWith(
        'click',
        expectedFunction
      );
    });
  });

  describe('toggle method', () => {
    it('change state', () => {
      const initialState = {
        isOpen: false
      };
      const wrapper = shallow(<HeaderContainer {...props} />);
      wrapper.setState(initialState);
      wrapper.instance().toggle();
      expect(wrapper.state().isOpen).toEqual(true);
    });
  });

  describe('handleDocumentClick method', () => {
    it('call toggle method if pass condition', () => {
      const e = {
        target: {
          id: 'email'
        }
      };
      const element = {
        contains: jest.fn(() => false)
      };
      const wrapper = shallow(<HeaderContainer {...props} />);
      const expectedFunction = (wrapper.instance().toggle = jest.fn());
      wrapper.instance().element = element;
      wrapper.instance().handleDocumentClick(e);
      expect(expectedFunction).toHaveBeenCalledTimes(1);
    });

    it('toggle method is not called', () => {
      const e = {
        target: {
          contains: jest.fn(() => true)
        }
      };
      const element = {
        contains: jest.fn(() => true)
      };
      const wrapper = shallow(<HeaderContainer {...props} />);
      const expectedFunction = (wrapper.instance().toggle = jest.fn());
      wrapper.instance().element = element;
      wrapper.instance().handleDocumentClick(e);
      expect(expectedFunction).toHaveBeenCalledTimes(0);
    });
  });

  describe('componentWillUnmount method', () => {
    it('call remove event', () => {
      const wrapper = shallow(<HeaderContainer {...props} />);
      wrapper.instance().componentWillUnmount();
      const expectedFunction = wrapper.instance().handleDocumentClick;
      expect(document.removeEventListener).toHaveBeenCalledWith(
        'click',
        expectedFunction
      );
    });
  });
});

describe('render', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<HeaderContainer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
