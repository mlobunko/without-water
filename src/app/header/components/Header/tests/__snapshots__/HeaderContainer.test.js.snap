// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
<div>
  <Component
    authorized={true}
    authorizedActionsLogOut={[Function]}
    isOpen={false}
    toggle={[Function]}
  />
</div>
`;
